doc/man3/SSL_CTX_set_cert_verify_callback.pod

```diff
@@ -20,24 +20,27 @@ the time when L<SSL_new(3)> is called.
 
 =head1 NOTES
 
-Whenever a certificate is verified during a SSL/TLS handshake, a verification
-function is called. If the application does not explicitly specify a
-verification callback function, the built-in verification function is used.
+When a peer certificate has been received during a SSL/TLS handshake,
+a verification function is called regardless of the verification mode.
+If the application does not explicitly specify a verification callback function,
+the built-in verification function is used.
 If a verification callback I<callback> is specified via
 SSL_CTX_set_cert_verify_callback(), the supplied callback function is called
-instead. By setting I<callback> to NULL, the default behaviour is restored.
-
-When the verification must be performed, I<callback> will be called with
-the arguments callback(X509_STORE_CTX *x509_store_ctx, void *arg). The
-argument I<arg> is specified by the application when setting I<callback>.
-
-I<callback> should return 1 to indicate verification success and 0 to
-indicate verification failure. If SSL_VERIFY_PEER is set and I<callback>
-returns 0, the handshake will fail. As the verification procedure may
-allow the connection to continue in the case of failure (by always
-returning 1) the verification result must be set in any case using the
-B<error> member of I<x509_store_ctx> so that the calling application
-will be informed about the detailed result of the verification procedure!
+instead with the arguments callback(X509_STORE_CTX *x509_store_ctx, void *arg).
+The argument I<arg> is specified by the application when setting I<callback>.
+By setting I<callback> to NULL, the default behaviour is restored.
+
+I<callback> should return 1 to indicate verification success
+and 0 to indicate verification failure.
+In server mode, a return value other than 1 leads to handshake failure.
+In client mode, the behaviour is as follows.
+A return value greater than 1 leads to handshake failure.
+Other values are ignored if the verification mode is B<SSL_VERIFY_NONE>.
+On return value 0 the handshake will fail.
+
+In any case the I<callback> should make sure that the verification result is set
+using L<X509_STORE_CTX_set_error(3)>. This way the calling application
+will be informed about the detailed result of the verification procedure.
```

Comment:
This is not an adequate description of the different roles of the callback's return value and the verification status value. The callback's return value is used only to optionally request that the handshake be aborted, this can happen even if the chain verification succeeds if for some reason the callback decides that the authenticated peer is `evil` and the connection should be aborted.

On the other hand the verification status in the store context is a possibly durable indication of the chain's validity! This gets recoded in the SSL session (and thus also in session tickets) and the validity of the originally presented chain is then visible on resumption, even though no chain is presented int that case.  (IIRC resumption preserves only the leaf certificate and its verification status, not the entire chain).

Thus returning the correct verification status is rather important, perhaps more so than whether the handshake is allowed to continue or not.  Indeed some applications always continue the handshake, but then gracefully disconnect at the application layer when peer authentication fails.

Reply:
> This is not an adequate description of the different roles of the callback's return value and the verification status value.

Note that I basically just moved the that paragraph while adding the hint that the verification result can be set using `X509_STORE_CTX_set_error()`.
Yet doing so I was slightly puzzled about what its author actually meant to say here.

> The callback's return value is used only to optionally request that the handshake be aborted, this can happen even if the chain verification succeeds

Well, it can also be used to overrule a non-successful verification (and the callback may not really do/invoke any verification).

> if for some reason the callback decides that the authenticated peer is `evil` and the connection should be aborted.

This reminds of the wonderful "Evil Bit" defined in RFC 3514 https://tools.ietf.org/html/rfc3514 :laughing: 

> On the other hand the verification status in the store context is a possibly durable indication of the chain's validity! This gets recorded in the SSL session (and thus also in session tickets) and the validity of the originally presented chain is then visible on resumption, even though no chain is presented int that case. (IIRC resumption preserves only the leaf certificate and its verification status, not the entire chain).

Interesting.

> Thus returning the correct verification status is rather important, perhaps more so than whether the handshake is allowed to continue or not. Indeed some applications always continue the handshake, but then gracefully disconnect at the application layer when peer authentication fails.

I see.

I've now tried to improve the paragraph according to the background you gave as follows.
> In any case a viable verification result value must be reflected
in the `B<error>` member of `I<x509_store_ctx>`,
which can be done using `L<X509_STORE_CTX_set_error(3)>`.
This is particularly important in case
the I<callback> allows the connection to continue (by returning 1).
Note that the verification status in the store context is a possibly durable
indication of the chain's validity!
This gets recorded in the SSL session (and thus also in session tickets)
and the validity of the originally presented chain is then visible
on resumption, even though no chain is presented int that case.
Moreover, the calling application will be informed about the detailed result of
the verification procedure any may want base further decisions on it.